src/deluge/deluge.cpp

```diff
@@ -324,42 +324,25 @@ bool readButtonsAndPads() {
 
 		if (value < PAD_AND_BUTTON_MESSAGES_END) {
 
-			int result;
-
 #if DELUGE_MODEL == DELUGE_MODEL_40_PAD
 			bool thisPadPressIsOn = (value >= 70);
-			int x = (unsigned int)value % 10;
-			int y = ((unsigned int)value % 70) / 10;
-
-			if (y < displayHeight) result = matrixDriver.padAction(x, y, thisPadPressIsOn, inSDRoutine);
-			else result = Buttons::buttonAction(x, y - displayHeight, thisPadPressIsOn, sdRoutineLock);
-
 #else
 			int thisPadPressIsOn = nextPadPressIsOn;
 			nextPadPressIsOn = USE_DEFAULT_VELOCITY;
+#endif
 
-			int y = (unsigned int)value / 9;
-			int x = value - y * 9;
-
-			// If pad...
-			if (y < displayHeight * 2) {
-				x <<= 1;
-				if (y >= displayHeight) {
-					y -= displayHeight;
-					x++;
-				}
+			int result;
+			if (MatrixDriver::isPad(value)) {
+				auto p = MatrixDriver::charToPad(value);
+				result = matrixDriver.padAction(p.x, p.y, thisPadPressIsOn);
```

Comment:
if you're feeling brave, the structure can be plumbed all the way down the call stack.

we should double check the generated assembly to make sure we're not spilling to RAM in that case, though I don't know how much it would really matter.